snapcraft/tests/test_parser.py

```diff
@@ -556,3 +561,41 @@ def test_missing_fields(self, mock_get, mock_get_origin_data):
         }
         main(['--debug', '--index', TEST_OUTPUT_PATH])
         self.assertEqual(0, _get_part_list_count())
+
+    def test__get_origin_data_both(self):
+        tempdir = tempfile.mkdtemp()
+        with open(os.path.join(tempdir, '.snapcraft.yaml'), 'w') as fp:
+            fp.write("")
+        with open(os.path.join(tempdir, 'snapcraft.yaml'), 'w') as fp:
+            fp.write("")
+
+        self.assertRaises(BadSnapcraftYAMLError, _get_origin_data, tempdir)
+        shutil.rmtree(tempdir)
+
+    def test__get_origin_data_hidden_only(self):
+        tempdir = tempfile.mkdtemp()
+        with open(os.path.join(tempdir, '.snapcraft.yaml'), 'w') as fp:
+            fp.write("")
+
+        _get_origin_data(tempdir)
+        shutil.rmtree(tempdir)
+
+    def test__get_origin_data_normal_only(self):
+        tempdir = tempfile.mkdtemp()
+        with open(os.path.join(tempdir, 'snapcraft.yaml'), 'w') as fp:
+            fp.write("")
+
+        _get_origin_data(tempdir)
+        shutil.rmtree(tempdir)
+
+    def test__encode_origin_git(self):
+        origin = 'git@github.com:testuser/testproject.git'
+        origin_dir = _encode_origin(origin)
+
+        self.assertEqual('gitgithub.comtestusertestproject.git', origin_dir)
+
+    def test__encode_origin_lp(self):
+        origin = 'lp:~testuser/testproject/testbranch'
+        origin_dir = _encode_origin(origin)
+
+        self.assertEqual('lptestusertestprojecttestbranch', origin_dir)
```

Comment:
This is not too bad, but ideally, we test only the public methods. In this case, all the tests should call run, not the internals. Did you try that approach? I understand it might require too much set up, or too many fakes. I'm not asking for it in this branch, but if the public method is not testable, that's a bug that should be reported.